### PR TITLE
[fparser2] Fixes for slow parsing and printing of code (#236)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
@reuterbal. Branch 236_slow_expressions should hopefully fix the reported problem in issue #236. It will probably take a little while for this PR to get on master as I will need to update the modified class and add appropriate tests. If you have time it would be good if you could check whether this does solve the problem for the full code, rather than just the test case you kindly provided.

**Comment by reuterbal** [comment]
I can confirm this fixes the reported issue, many thanks!

Looking at the call graph the reduction in matches is well visible: Went from more than 4 million calls to `match` down to <5k, and invocations of `Base::__new__` reduced from more than 13 million down to 16.2k.
fparser-issue-fixed

Overall parse time reduced from multiple minutes to <0.5sec on the example as well as the full source files that caused problems before.

I have tested the full code base using your branch and overall parse time stays roughly the same. Comparing 0.0.10 from PyPi, master and your branch, the latter two showed a slight slowdown (~2-3%) compared to 0.0.10. But this was just a quick test and may well be some OS hiccup or similar things.

Now it is possible to parse the entire source tree including the two problematic files that were skipped previously. Support for directives and exceptions thrown by few other things we'll look at next.

Again, thanks a lot!

**Comment by rupertford** [comment]
Great. It's likely you will find some other issues and if so please feel free to feed those back too.

**Comment by reuterbal** [comment]
Absolutely, we will keep you in the loop!

**Comment by reuterbal** [comment]
This has been stale for a while. Anything I can do to help?
Currently it requires a small fix (https://github.com/reuterbal/fparser/commit/5663bfd24e0c86ed9b38d950726ab101f76c22a6) but with that it works fine for us.

**Comment by rupertford** [comment]
Again, thanks for the prompt. I'll get this ready for review and, if you are happy, incorporate your fixes - although I would hope they would come out in the tests.

**Comment by reuterbal** [comment]
Thanks for the reply, I hope you didn't mind me nudging (was rebasing our integration branch and thought it wouldn't hurt).

My fixes are only changing return types in `BinaryOpBase.match` back to `return None` from `return False`. Tests are actually failing without that and pass again when doing this.
(The reason this shows up as multiple commits in the PR is because from time to time I rebase my integration branch to master and merge the open PRs that I need. Cherry-picking the commit with the fixes, results in a new hash that shows then up again every time.)

File: src/fparser/two/utils.py [hunk 1]
```diff
@@ -874,49 +874,128 @@ def match(op_pattern, rhs_cls, string, exclude_op_pattern=None):
...
+        :param exclude_op_pattern: optional argument which specifies a \
+            particular subpattern to exclude from the match. Defaults \
+            to None which means there is no subpattern.
+        :type exclude_op_pattern: :py:class:`fparser.two.pattern_tools.Pattern`
+        :param bool is_add: optional match optimisation for the + \
+            operator when set to True (ignores matching the + in a
+            real literal). Defaults to False.
+
+        :returns: a tuple containing the matched lhs, the operator and \
+            the matched rhs of the input string or None is there is \
```
Comment (arporter) [inline src/fparser/two/utils.py:941]: s/is there is/if there is/

File: src/fparser/two/utils.py [hunk 2]
```diff
@@ -874,49 +874,128 @@ def match(op_pattern, rhs_cls, string, exclude_op_pattern=None):
 
 
 class BinaryOpBase(Base):
```
Comment (arporter) [inline src/fparser/two/utils.py:876]: Please update the copyright date on this file (and the other one if I forgot to say that).

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good - thanks for re-working the match routine.
I only found a few tiny things to fix.
Once they're done, I shall merge.

File: src/fparser/two/tests/utils/test_binaryopbase.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+# This is required to setup the fortran2003 classes (when matching
+# with Complex_Literal_Constant)
+_ = ParserFactory().create(std="f2003")
+
+
+@pytest.mark.parametrize('pattern', ["%", fparser_patterns.percent_op])
+@pytest.mark.parametrize('right', [True, False])
+def test_binaryopbase_pattern_nomatch(pattern, right):
+    '''Test the BinaryOpBase match method returns None if the pattern is
+    of type 'str' or 'pattern' and is not found in the string. Check
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/utils/test_binaryopbase.py:55]: Yes, thanks.

File: src/fparser/two/tests/utils/test_binaryopbase.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        _ = BinaryOpBase.match(lhs, pattern, rhs, string, right=right)
+    assert "Name: '{0}'".format(expected) in str(info.value)
+
+
+@pytest.mark.parametrize('pattern', ["%", fparser_patterns.percent_op.named()])
+@pytest.mark.parametrize('string', ["a%b", "  a  %  b  "])
+def test_binaryopbase_match(pattern, string):
+    '''Test the BinaryOpBase match method returns the expected results if
+    there is a match for a pattern of type str and Pattern. Also check
+    that spaces no not affect matching and that they are removed in
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/utils/test_binaryopbase.py:111]: Oops.

File: src/fparser/two/utils.py (hunk 1 above)
Comment (rupertford) [reply src/fparser/two/utils.py:941]: Oops again.

File: src/fparser/two/utils.py (hunk 2 above)
Comment (rupertford) [reply src/fparser/two/utils.py:876]: Updated.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — APPROVED** [review]

**Pull request merged**